File: lintcode/Longest_Common_Substring.cpp
class Solution {
public:    
    /**
     * @param A, B: Two string.
     * @return: the length of the longest common substring.
     */
    int longestCommonSubstring(string &A, string &B) {
        // write your code here
        int m = A.size();
        int n = B.size();
        
        vector<vector<int>> tbl(m, vector<int>(n, 0));
        int maxLen = 0;
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                 if(i == 0 || j == 0) tbl[i][j] = (A[i] == B[j])? 1 : 0;
                 
                 if(i > 0 && j > 0){
                     if(A[i] == B[j]) tbl[i][j] = tbl[i-1][j-1] +1;
                     else tbl[i][j] = 0;
                 }
                 
                 if(tbl[i][j] > maxLen) maxLen = tbl[i][j];
            }
        }
        
        return maxLen;
    }
};

File: lintcode/java/LongestCommonSubstring.java
public class Solution {
    /**
     * @param A, B: Two string.
     * @return: the length of the longest common substring.
     */
    public int longestCommonSubstring(String A, String B) {
        
        int m = A.length();
        int n = B.length();
        int[][] t = new int[m+1][n+1];
        
        int maxLen = 0;
        for(int i = 1; i <= m; i++){
            for(int j = 1; j <= n; j++){
                if(A.charAt(i-1) == B.charAt(j-1)){
                    t[i][j] = t[i-1][j-1] + 1;
                    if(t[i][j] > maxLen) maxLen = t[i][j];
                } else {
                    t[i][j] = 0;
                }
            }
        }
        
        return maxLen;
    }
}
